File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output = []
        numZeros = 0
        prod = 1
        for num in nums:
            if num == 0:
                numZeros += 1
            else:
                prod *= num
          
        for num in nums:
            if num == 0:
                output.append(0) if numZeros > 1 else output.append(int(prod))   
            else:
                output.append(0) if numZeros > 0 else output.append(int(prod*(num**-1)))
                    
        return output

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prefix = 1;
        vector<int> result (nums.size(), 0);
        
        for (int i = nums.size() - 1; i >= 0; --i) {
            if (i < nums.size() - 1) {
                result[i] = nums[i] * result[i + 1];
            } else {
                result[i] = nums[i];
            }
        }

        for (int i = 0; i < nums.size(); ++i) {
            if (i < nums.size() - 1) {
                result[i] = prefix * result[i + 1];
            } else {
                result[i] = prefix;
            }
            prefix *= nums[i];
        }

        return result;
    }
};